Upgrade.md

```diff
@@ -0,0 +1,86 @@
+# Upgrade
+
+All instructions to  upgrade this project from  one release to the  next will be
+documented in this  file. Upgrades must be run sequentially,  meaning you should
+not skip minor/major releases while upgrading (fix releases can be skipped).
+
+The format is inspired from [Keep a
+Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to
+[Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+
+## Unreleased
+
+### Static services
+
+This version introduces `Static services` and it is a breaking change in the
+blue/green deployment process. The `switch` playbook is patching static services
+instead of routes. And the `deployment_stamp` of the deployed stacks is no
+longer stored in routes, but in static services instead.
+
+In order to migrate to this new system, you have to apply the following upgrade
+process. This will not cause any downtime, but it implies a redeployment of your
+applications.
+
+#### Upgrade Procedure
+
+This procedure must be applied for every arnold application deployed.
+
+The `current` stack of your application (before applying this procedure) will be
+referenced as `BEFORE-UPGRADE-VERSION`.
+
+
+##### 1. Upgrade your arnold version
+
+##### 2. Execute the `create_services` ansible playbook :
+
+```
+bin/ansible-playbook create_services.yml -e "apps_filter=your-application"
+```
+It will create the static  services required by your application. Each variant of these static services (`previous`, `current`, `next`) willtarget nothing after this step.
+
+##### 3. Execute the `deploy` ansible playbook :
+
+```
+bin/deploy -e "apps_filter=your-application"
+```
+
+It will deploy your application and the `next` static services will target this
+stack.
+
+This step does not affect routes. So your `BEFORE-UPGRADE-VERSION` stays
+unchanged and available via the `current` route. On the other hand, the `next`
+stack you just deployed will not be reachable by any route.
+
+##### 4. Execute the `switch` ansible playbook :
+
+```
+bin/switch -e "apps_filter=your-application"
+```
+
+The switch will be done at a the static services level. So, the stack you just
+deployed will be considered as the `current` by static services and by the
+ansible playbooks.
+
+But again, this step does not affect routes. So your `BEFORE-UPGRADE-VERSION`
+stack stays unchanged and available via the `current` route.
+
+##### 5. Execute the `create_routes` playbook :
+
+```
+bin/ansible-playbook create_routes.yml -e "apps_filter=your-application"
```

Comment:
are you sure that this will not erase all SSL certificates?

Reply:
The routes already exist, so this playbook will only patch them with `deploy_patch_route.yml` to change the name of the targeted service, just like with the `switch_route.yml` task on master branch. So I assume nothing will be lost, right ?

Reply:
The last time I ran create_routes it completely reset all my routes. This has to be tested to be sure.

Reply:
Yes, I can explain this behavior : when you execute the `create_routes` playbook of the master branch, it does not retrieve the deployment_stamp of the deployed stacks. As a result of what, the routes are patched to target an undefined service (i.e. route "hello-current" targets the service "hello-") and are broken.

The version of `create_routes` in this branch is simplified because it always points to a static service (i.e. route "hello-current" always targets service "hello-current").

I successfully tested the upgrade process on my local openshift cluster with the hello app.
The routes are patched and not re-created, so there should be no loss related to TLS certificates.

But of course we have to test it in a non-production environment.